crates/isograph_lsp/src/goto_definition.rs

```diff
@@ -0,0 +1,181 @@
+use lsp_types::{Position, Uri};
+use pico_macros::memo;
+use resolve_position::ResolvePosition;
+use std::borrow::Cow;
+use std::ops::Deref;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+pub fn on_goto_definition<TNetworkProtocol: NetworkProtocol + 'static>(
+    compiler_state: &CompilerState<TNetworkProtocol>,
+    params: <GotoDefinition as Request>::Params,
+) -> LSPRuntimeResult<<GotoDefinition as Request>::Result> {
+    let db = &compiler_state.db;
+    Ok(on_goto_definition_impl(
+        db,
+        params.text_document_position_params.text_document.uri,
+        params.text_document_position_params.position,
+    )
+    .to_owned()?
+    .map(GotoDefinitionResponse::Scalar))
+}
+
+#[memo]
+pub fn on_goto_definition_impl<TNetworkProtocol: NetworkProtocol + 'static>(
+    db: &IsographDatabase<TNetworkProtocol>,
+    url: Uri,
+    position: Position,
+) -> LSPRuntimeResult<Option<lsp_types::Location>> {
+    let current_working_directory = db.get_current_working_directory();
+
+    let relative_path_to_source_file = relative_path_from_absolute_and_working_directory(
+        current_working_directory,
+        &url.to_file_path().expect("Expected file path to be valid."),
+    );
+
+    let extraction_option =
+        get_iso_literal_extraction_from_text_position_params(db, url.clone(), position.into())
+            .to_owned();
+    let (extraction, offset) = match extraction_option {
+        Some(e) => e,
+        None => return Ok(None),
+    };
+
+    let memo_ref = get_validated_schema(db);
+    let (validated_schema, _stats) = match memo_ref.deref() {
+        Ok(schema) => schema,
+        Err(_) => return Ok(None),
+    };
+
+    let goto_location = if let Ok((result, _text_source)) = process_iso_literal_extraction(
+        db,
+        &extraction,
+        relative_path_to_source_file,
+        current_working_directory,
+    ) {
+        match result.resolve((), Span::new(offset, offset)) {
+            IsographResolvedNode::ClientFieldDeclaration(_) => None,
+            IsographResolvedNode::ClientPointerDeclaration(_) => None,
+            IsographResolvedNode::EntrypointDeclaration(_) => None,
+            IsographResolvedNode::ServerObjectEntityNameWrapper(_) => None,
+            IsographResolvedNode::Description(_) => None,
+            IsographResolvedNode::ScalarSelection(scalar_path) => {
+                if let Ok((_, selectable)) =
+                    get_parent_and_selectable_for_scalar_path(&scalar_path, validated_schema)
+                {
+                    match selectable {
+                        DefinitionLocation::Server(server_selectable) => {
+                            server_definition_location(db, &server_selectable.name.location)
+                                .to_owned()
+                        }
+                        DefinitionLocation::Client(client_selectable) => {
+                            server_definition_location(db, &client_selectable.name.location)
```

Comment:
lets rename that fn, since it's not just for server definitions, and put it somewhere else. It's not specific to goto def (e.g. find all references will use it, maybe we can incorporate this in hover.)